app/utils/asyncValidator.js

```diff
@@ -0,0 +1,53 @@
+// @flow
+
+export type AsyncFieldValidator<T = any> = (
+  fieldValue: T,
+  formData: Object
+) => Promise<[true] | [boolean, string]>;
+
+type FieldValidators = {|
+  [field: string]: AsyncFieldValidator<>[],
+|};
+
+const getFieldErrors = async (
+  validators: AsyncFieldValidator<>[],
+  field: string,
+  formData: Object
+): Promise<[string, string[]]> => {
+  const validationResults = await Promise.all(
+    validators.map((validator) => validator(formData[field], formData))
+  );
+
+  const errors = validationResults
+    .filter(([isValid]) => !isValid)
+    // $FlowFixMe flow doesn't understand that the filter ensures that message exists
+    .map(([, error]) => error);
+
+  return [field, errors];
+};
+
+const getFieldErrorArray = (
+  fieldValidators: FieldValidators,
+  formData: Object
+): Promise<[string, string[]][]> => {
+  return Promise.all(
+    Object.keys(fieldValidators).map((field) =>
+      getFieldErrors(fieldValidators[field], field, formData)
+    )
+  ).then((fieldArray) =>
+    fieldArray.filter(([, fieldErrors]) => fieldErrors.length)
+  );
+};
```

Comment:
I agree that it is generally better with less code, yet functions should do one thing and be rather short. While this is not long, it contains a log of logic and chained calls, so IMO it is more readable with a helper functions.

Reply:
To quote the linux kernel coding style

> However, if you have a complex function, and you suspect that a less-than-gifted first-year high-school student might not even understand what the function is all about, you should adhere to the maximum limits all the more closely. Use helper functions with descriptive names

Reply:
I generally agree, but when the helper functions are < 10 lines and it's _not_ easy to understand what it does, it could have the opposite effect. I don't necessarily mean the example above is the best. As I said, maybe just merging the two helper functions into one will help.

Generally this is a JS issue though :shrug:, with a mix of chaining, function calls and Promise/awaits it becomes ugly no matter how you do it.

Reply:
I'm a big fan of splitting functions into simpler helper functions to increase readability. In this case I don't think it is very easily readable, it's a bit tricky because of the Promises and nested mapping, but putting it all in a single function makes it even less readable imo. even though it does use a lot fewer lines of code.